File: src/main/java/tech/macil/minecraft/rcon/RconPlugin.java
package tech.macil.minecraft.rcon;

import com.google.common.base.Charsets;
import org.apache.commons.lang.exception.ExceptionUtils;
import org.bukkit.command.CommandSender;
import org.bukkit.plugin.java.JavaPlugin;
import tech.macil.util.NLRequiringBufferedReader;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketException;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.logging.Level;

public class RconPlugin extends JavaPlugin {
    private static final int SOCKET_BACKLOG = 20;
    private static final int THREAD_COUNT = 6;
    private static final String EXPECTED_GREETING = "Minecraft-Rcon";
    private ServerSocket socket;
    private static final Executor connectionHandler = Executors.newWorkStealingPool(THREAD_COUNT);
    private static final Executor outputFlusher = Executors.newWorkStealingPool(THREAD_COUNT);

    @Override
    public void onEnable() {
        saveDefaultConfig();

        String listenAddress = getConfig().getString("listenAddress");
        int port = getConfig().getInt("port");

        try {
            if (listenAddress.equals("all")) {
                socket = new ServerSocket(port, SOCKET_BACKLOG);
            } else {
                socket = new ServerSocket(port, SOCKET_BACKLOG, InetAddress.getByName(listenAddress));
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        new Thread(() -> {
            try {
                while (true) {
                    Socket connection = socket.accept();
                    connectionHandler.execute(new ClientConnectionRunnable(connection));
                }
            } catch (SocketException e) {
                // ignore, happens when socket is closed
            } catch (IOException e) {
                getLogger().log(Level.SEVERE, "Unknown error in server thread", e);
            }
        }).start();
    }

    @Override
    public void onDisable() {
        if (socket != null) {
            try {
                socket.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            } finally {
                socket = null;
            }
        }
    }

    private class ClientConnectionRunnable implements Runnable {
        private final Socket connection;

        ClientConnectionRunnable(Socket connection) {
            this.connection = connection;
        }

        @Override
        public void run() {
            try {
                try {
                    // Use NLRequiringBufferedReader so if the connection dies part way through, we don't
                    // execute half of a command.
                    BufferedReader input = new NLRequiringBufferedReader(
                            new InputStreamReader(connection.getInputStream(), Charsets.UTF_8));

                    try (OffThreadWriter output = new OffThreadWriter(
                            new PrintWriter(connection.getOutputStream(), false),
                            getLogger(),
                            outputFlusher
                    )) {
                        CommandSender sender = new RconCommandSender(RconPlugin.this, output);

                        String greeting = input.readLine();
                        if (!EXPECTED_GREETING.equals(greeting)) {
                            return;
                        }

                        input.lines().forEach(line -> {
                            getLogger().log(Level.INFO, "rcon(" + connection.getRemoteSocketAddress() + "): " + line);
                            try {
                                if (getServer().getScheduler().callSyncMethod(RconPlugin.this, () ->
                                        !getServer().dispatchCommand(sender, line)
                                ).get()) {
                                    output.writeLnWithoutFlush("Command not found");
                                    output.flush();
                                }
                            } catch (Exception e) {
                                output.writeLnWithoutFlush(ExceptionUtils.getStackTrace(e));
                                output.flush();
                            }
                        });
                    }
                } finally {
                    connection.close();
                }
            } catch (Exception e) {
                getLogger().log(Level.SEVERE, "Unknown error in connection thread", e);
            }
        }
    }
}

File: src/main/kotlin/tech/macil/minecraft/rcon/RconPlugin.kt
package tech.macil.minecraft.rcon

import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.core.Logger
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import tech.macil.minecraft.rcon.util.BlockerCounter
import tech.macil.minecraft.rcon.util.ByteArrayQueue

import java.io.*
import java.time.Duration
import java.util.concurrent.CancellationException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.logging.*
import javax.servlet.ServletOutputStream
import kotlin.concurrent.thread

class RconPlugin : JavaPlugin() {
    private var webServer: WebServer? = null
    private val disablingCf = CompletableFuture<Unit>()
    private val runningRequestsCounter = BlockerCounter()

    override fun onEnable() {
        saveDefaultConfig()

        Metrics(this)

        var listenAddress: String? = config.getString("listenAddress")
        if (listenAddress == "all") listenAddress = null
        val port = config.getInt("port")

        val htpasswdFile = File(dataFolder, "htpasswd")
        val authChecker: AuthChecker? = if (htpasswdFile.exists()) {
            logger.log(Level.INFO, "Found htpasswd file, loading it.")
            AuthChecker(htpasswdFile)
        } else {
            null
        }

        val webServer = WebServer(listenAddress, port, this, authChecker, this::handleCommand)
        this.webServer = webServer

        webServer.start()
    }

    override fun onDisable() {
        // Signal to any current requests that the plugin is shutting down.
        disablingCf.complete(null)

        // Wait on any current requests to finish before killing the webserver.
        runningRequestsCounter.await(Duration.ofSeconds(10).toMillis())

        val webServer = this.webServer
        if (webServer != null) {
            this.webServer = null
            webServer.stop()
        }
    }

    private fun handleCommand(command: String, output: ServletOutputStream, remoteAddr: String) {
        runningRequestsCounter.acquire().use {
            logger.log(Level.INFO, "rcon($remoteAddr): $command")

            val (queuedInput, queuedOutput) = ByteArrayQueue.makePair()

            val writerThread = thread {
                output.use {
                    queuedInput.use {
                        val buffer = ByteArray(4096)
                        while (true) {
                            val len = queuedInput.read(buffer)
                            if (len < 0) break
                            output.write(buffer, 0, len)
                        }
                    }
                }
            }

            PrintWriter(queuedOutput).use { queuedPrintWriter ->
                val appender = RconAppender(queuedPrintWriter)
                try {
                    (LogManager.getRootLogger() as Logger).addAppender(appender)
                    try {
                        val dispatchCf = CompletableFuture<Unit>()
                        server.scheduler.callSyncMethod(this) {
                            server.dispatchCommand(Bukkit.getConsoleSender(), command)
                            dispatchCf.complete(null)
                        }
                        // If the command causes this plugin to unload, then dispatchCommand()
                        // will call into onDisable() which will call webServer.stop() which
                        // will wait on this thread to end, so we need to bail on waiting on
                        // the dispatchCf future if onDisable() is called.
                        CompletableFuture.anyOf(dispatchCf, disablingCf).get()

                        waitForLogsToFlush()
                    } catch (e: Exception) {
                        e.printStackTrace(queuedPrintWriter)
                    }
                } finally {
                    (LogManager.getRootLogger() as Logger).removeAppender(appender)
                }
            }
            writerThread.join()
        }
    }

    private fun waitForLogsToFlush() {
        if (disablingCf.isDone) return

        // It seems like the output of some commands is only delivered to our appender
        // asynchronously. I couldn't find a direct way to wait on whatever loggers involved
        // to flush, but I found that just waiting on the next game tick seems to (maybe just
        // mostly) solve the problem in practice.

        // There's a similar issue that some plugins' commands (especially any plugins using a
        // database) don't output any results until some unknown time later. This doesn't help
        // much for those, and I'm not sure if I intend to address that.

        // The timeout is because when the server is shutting down, the scheduled method never
        // runs.
        try {
            server.scheduler.callSyncMethod(this) { null }.get(500, TimeUnit.MILLISECONDS)
        } catch (e: CancellationException) {
            // pass
        } catch (e: TimeoutException) {
            // pass
        }
    }
}
